### PR TITLE
Added Command and Config modules to support Dell Networking OS10 device

### PR DESCRIPTION
**Pull request by skg-net**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME
- modules/core/network/dnos10_command
- modules/core/network/dnos10_config
##### ANSIBLE VERSION

ansible --version
ansible 2.2.0 (devel 35a3653) last updated 2016/08/25 12:06:08 (GMT -700)
lib/ansible/modules/core: (detached HEAD ef84dbb) last updated 2016/08/25 12:06:38 (GMT -700)
lib/ansible/modules/extras: (detached HEAD f29efb5626) last updated 2016/08/25 12:06:38 (GMT -700)
config file =
configured module search path = Default w/o overrides
##### SUMMARY
- Added dnos10_command module and dnos10_config module, to support DELL Networking OS10 devices.

### CODE REVIEW

**Comment by skg-net** [comment]
**Ansible Repo PR:**
[https://github.com/ansible/ansible/pull/17296](url)

**Test Modules PR:**
[https://github.com/ansible/test-network-modules/pull/2](url)

**Comment by gundalow** [comment]
Please add an `__init__.py`file in the new `dell` directory

**Comment by gregdek** [comment]
Thanks @skg-net for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

File: network/dell/dnos10_command.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+def to_lines(stdout):
+    for item in stdout:
+        if isinstance(item, basestring):
+            item = str(item).split('\n')
+        yield item
+
+def main():
+    spec = dict(
+        commands=dict(type='list', required=True),
+        wait_for=dict(type='list', aliases=['waitfor']),
```
Comment (gundalow) [inline network/dell/dnos10_command.py]: @privateip Do you think we need to add the alias for `waitfor` on new modules?
I'm in two minds, 
On one hand it keeps it consistent across Networking Modules, on the other hand, this is a new module and therefore doesn't have any legacy issues

File: network/dell/dnos10_config.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+#
+
+DOCUMENTATION = """
+---
+module: dnos10_config
+version_added: "2.2"
+author: "Senthil Kumar Ganesan (@skg_net)"
+short_description: Manage Dell OS10 configuration sections
+description:
+  - Dell OS10 configurations use a simple block indent file sytanx
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/sytanx/syntax/

File: network/dell/dnos10_config.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    for segementing configuration into sections.  This module provides
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/segementing/segmenting/

File: network/dell/dnos10_config.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    a deterministic way.
+extends_documentation_fragment: dnos10
+options:
+  lines:
+    description:
+      - The ordered set of commands that should be configured in the
+        section.  The commands must be the exact same commands as found
+        in the device running-config.  Be sure to note the configuration
+        command syntax as some commands are automatically modified by the
+        device config parser.
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: Please add: 
`This argument is mutually exclusive with O(src).`

File: network/dell/dnos10_config.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        the modified lines are pushed to the device in configuration
+        mode.  If the replace argument is set to I(block) then the entire
+        command block is pushed to the device in configuration mode if any
+        line is not correct
+    required: false
+    default: line
+    choices: ['line', 'block']
+  update_config:
+    description:
+      - This arugment will either cause or prevent the changed commands
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/arugment/argument/

File: network/dell/dnos10_config.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        exclusive with O(lines).
+    required: false
+    default: null
+  before:
+    description:
+      - The ordered set of commands to push on to the command stack if
+        a change needs to be made.  This allows the playbook designer
+        the opportunity to perform configuration commands prior to pushing
+        any changes without affecting how the set of commands are matched
+        against the system
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: Please add missing full stop at the end of the sentence.

File: network/dell/dnos10_command.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+#
+
+DOCUMENTATION = """
+---
+module: dnos10_command
+version_added: "2.2"
+author: "Senthil Kumar Ganesan (@skg_net)"
+short_description: Run commands on remote devices running Dell OS10
+description:
+  - Sends arbitrary commands to an Dell OS10 node and returns the results
```
Comment (gundalow) [inline network/dell/dnos10_command.py]: s/to an Dell/to a Dell/

File: network/dell/dnos10_command.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    Please use M(dnos10_config) to configure Dell OS10 devices.
+extends_documentation_fragment: dnos10
+options:
+  commands:
+    description:
+      - List of commands to send to the remote dnos10 device over the
+        configured provider. The resulting output from the command
+        is returned. If the I(wait_for) argument is provided, the
+        module is not returned until the condition is satisfied or
+        the number of retries has expired.
```
Comment (gundalow) [inline network/dell/dnos10_command.py]: `I(retries)`

File: network/dell/dnos10_config.py (hunk 2 above)
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/sytanx/syntax/

File: network/dell/dnos10_config.py (hunk 3 above)
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/segementing/segmenting/

File: network/dell/dnos10_config.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        command syntax as some commands are automatically modified by the
+        device config parser.
+    required: false
+    default: null
+    aliases: ['commands']
+  parents:
+    description:
+      - The ordered set of parents that uniquely identify the section
+        the commands should be checked against.  If the parents argument
+        is omitted, the commands are checked against the set of top
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: What decides if the commands are checked against "top level" or "global commands"?

File: network/dell/dnos10_config.py (hunk 4 above)
Comment (gundalow) [inline network/dell/dnos10_config.py]: Please add

`This argument is mutually  exclusive with O(src).`

File: network/dell/dnos10_config.py (hunk 6 above)
Comment (gundalow) [inline network/dell/dnos10_config.py]: Please add missing full stop at the end of this sentence.

File: network/dell/dnos10_config.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    default: line
+    choices: ['line', 'strict', 'exact', 'none']
+  replace:
+    description:
+      - Instructs the module on the way to perform the configuration
+        on the device.  If the replace argument is set to I(line) then
+        the modified lines are pushed to the device in configuration
+        mode.  If the replace argument is set to I(block) then the entire
+        command block is pushed to the device in configuration mode if any
+        line is not correct
```
Comment (gundalow) [inline network/dell/dnos10_config.py]: Please add missing full stop at the end of this sentence.
 Add a line note

File: network/dell/dnos10_config.py (hunk 5 above)
Comment (gundalow) [inline network/dell/dnos10_config.py]: s/arugment/argument/

**Comment by gundalow** [comment]
needs_revision

I'll let @privateip do the code review side.

Note, Travis will fail with `ERROR: DOCUMENTATION fragment missing: dnos10` till https://github.com/ansible/ansible/pull/17296 is merged in, we will get that is as a priority.

File: network/dell/dnos10_command.py (hunk 1 above)
Comment (privateip) [reply network/dell/dnos10_command.py]: Im okay with no alias but the doc string needs to match the argument_spec (which it doesnt right now)

File: network/dell/dnos10_config.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+        before=dict(type='list'),
+        after=dict(type='list'),
+
+        match=dict(default='line',
+                   choices=['line', 'strict', 'exact', 'none']),
+        replace=dict(default='line', choices=['line', 'block']),
+
+        update_config=dict(type='bool', default=True),
+        backup_config=dict(type='bool', default=False)
```
Comment (privateip) [inline network/dell/dnos10_config.py]: to remain consistent with other network *_config modules please rename argument to `backup`.  You can add an alias to `backup_config` if you want to use it

File: network/dell/dnos10_config.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+            result['responses'] = response
+
+            if module.params['save_config']:
+                module.config.save_config()
+                result['saved'] = True
+
+        result['changed'] = True
+
+    result['updates'] = commands
+    result['connected'] = module.connected
```
Comment (privateip) [inline network/dell/dnos10_config.py]: not needed.  You can leave here if you want to but it needs to be documented in the RETURNS doc string

File: network/dell/dnos10_config.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+                           supports_check_mode=True)
+
+    module.check_mode = not module.params['update_config']
+
+    parents = module.params['parents'] or list()
+
+    match = module.params['match']
+    replace = module.params['replace']
+    before = module.params['before']
+    result = dict(changed=False, saved=False)
```
Comment (privateip) [inline network/dell/dnos10_config.py]: The return value `saved` needs to be documented in the RETURNS string

File: network/dell/dnos10_config.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+        src=dict(type='path'),
+
+        before=dict(type='list'),
+        after=dict(type='list'),
+
+        match=dict(default='line',
+                   choices=['line', 'strict', 'exact', 'none']),
+        replace=dict(default='line', choices=['line', 'block']),
+
+        update_config=dict(type='bool', default=True),
```
Comment (privateip) [inline network/dell/dnos10_config.py]: The `update_config` argument was changed to `update` and is now a list of choices in other *_config modules if trying to remain consistent.  I would highly recommend changing this to `update = dict(default='merge', choices=['merge', 'check'])`

File: network/dell/dnos10_config.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        the current C(running-config) from the remote device before any
+        changes are made.  The backup file is written to the C(backup)
+        folder in the playbook root directory.  If the directory does not
+        exist, it is created.
+    required: false
+    default: no
+    choices: ['yes', 'no']
+"""
+
+EXAMPLES = """
```
Comment (privateip) [inline network/dell/dnos10_config.py]: Add required `provider` argument to EXAMPLES.  See [here](https://github.com/ansible/ansible-modules-core/blob/devel/network/ios/ios_config.py#L189) for sample

File: network/dell/dnos10_config.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+    argument_spec.update(dnos10_argument_spec)
+
+    mutually_exclusive = [('lines', 'src')]
+
+    module = NetworkModule(argument_spec=argument_spec,
+                           connect_on_load=False,
+                           mutually_exclusive=mutually_exclusive,
+                           supports_check_mode=True)
+
+    module.check_mode = not module.params['update_config']
```
Comment (privateip) [inline network/dell/dnos10_config.py]: remove based on comment above about use of the argument `update`

**Comment by privateip** [comment]
some general comments
- we can work through this PR but for future reference please make on PR per module
- please change the folder name to dnos10 from dell.  We categorize modules based on OS name

**Comment by privateip** [comment]
dns10_command looks good which the few doc string corrections referenced by gundalow

**Comment by privateip** [comment]
dns10_config needs some work.  It looks like you are trying to maintain consistency with the other *_config modules which is good as it makes it much easier on the playbook designers and facilitates easier role creation.  However, I think you are loosely basing it on an older *_config implementation.   I would recommend reviewing the implementation [here](https://github.com/ansible/ansible-modules-core/blob/devel/network/ios/ios_config.py) and making corrections as needed if that is the case.

File: network/dell/dnos10_command.py (hunk 1 above)
Comment (skg-net) [reply network/dell/dnos10_command.py]: Removed the alias for wait_for

File: network/dell/dnos10_config.py (hunk 9 above)
Comment (skg-net) [reply network/dell/dnos10_config.py]: None, currently it checks against both the global and top level commands.

**Comment by skg-net** [comment]
@privateip I have incorporated your  review comments for dnos10_config, but it still follows some of the  old implementation. Once this is in, I can work to align with the latest implementation.

**Comment by skg-net** [comment]
ready_for_review

**Comment by privateip** [comment]
looks good.   travis errors are due to ansible/ansible#17296 not being merged yet.  will merge these as soon as we get 17296 merged

**Comment by gregdek** [comment]
Thanks @skg-net for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by skg-net** [comment]
ready_for_review

Addressed @Qalthos comments from PR # https://github.com/ansible/ansible/pull/17296

**Comment by gregdek** [comment]
Thanks @skg-net for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by skg-net** [comment]
ready_for_review

**Pull request merged**